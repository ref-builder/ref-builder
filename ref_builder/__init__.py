"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
