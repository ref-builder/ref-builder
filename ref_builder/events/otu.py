from pydantic import UUID4

from ref_builder.events.base import EventData, Event, OTUQuery
from ref_builder.models import Molecule
from ref_builder.plan import MonopartitePlan, MultipartitePlan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: MonopartitePlan | MultipartitePlan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery


class ExcludeAccessionData(EventData):
    """The data for a :class:`ExcludeAccession` event."""

    accession: str


class ExcludeAccession(Event):
    """An accession exclusion event.

    This event is emitted when a Genbank accession is not going to be allowed in the
    reference.
    """

    data: ExcludeAccessionData
    query: OTUQuery


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: MonopartitePlan | MultipartitePlan


class CreatePlan(Event):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery


class SetReprIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetReprIsolate(Event):
    """An event that sets the representative isolate for an OTU."""

    data: SetReprIsolateData
    query: OTUQuery
