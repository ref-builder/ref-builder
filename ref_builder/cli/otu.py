import sys
from pathlib import Path
from uuid import UUID

import click
from click import Context
import structlog

from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.console import print_otu, print_otu_list
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.create import create_otu
from ref_builder.otu.isolate import (
    add_unnamed_isolate,
    add_and_name_isolate,
    add_genbank_isolate,
)
from ref_builder.otu.modify import (
    exclude_accessions_from_otu,
    set_representative_isolate,
    add_segments_to_plan,
    resize_monopartite_plan,
)
from ref_builder.otu.update import auto_update_otu, promote_otu_accessions
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.plan import SegmentRule, SegmentName
from ref_builder.repo import Repo
from ref_builder.utils import IsolateNameType, IsolateName

logger = structlog.get_logger()


@click.group(name="otu")
def otu() -> None:
    """Manage OTUs."""


@otu.command(name="create")
@click.argument("TAXID", type=int)
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--acronym", type=str, default="")
@ignore_cache_option
@path_option
def otu_create(
    accessions_: list[str],
    acronym: str,
    ignore_cache: bool,
    path: Path,
    taxid: int,
) -> None:
    """Create a new OTU.

    OTUs are created from a list of accessions and a taxonomy ID. The associated Genbank
    records are fetched and used to create the first isolate and a plan.
    """
    repo = Repo(path)

    if len(accessions_) != len(set(accessions_)):
        click.echo("Duplicate accessions were provided.", err=True)
        sys.exit(1)

    try:
        create_otu(
            repo,
            taxid,
            accessions_,
            acronym=acronym,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@otu.command(name="get")
@click.argument("IDENTIFIER", type=str)
@path_option
def otu_get(identifier: str, path: Path) -> None:
    """Get an OTU by its unique ID or taxonomy ID."""
    try:
        identifier = int(identifier)
        otu_ = Repo(path).get_otu_by_taxid(identifier)
    except ValueError:
        otu_ = Repo(path).get_otu(UUID(identifier))

    if otu_ is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    print_otu(otu_)


@otu.command(name="list")
@path_option
def otu_list(path: Path) -> None:
    """List all OTUs in the repository."""
    print_otu_list(Repo(path).iter_minimal_otus())


@otu.group(invoke_without_command=True)
@click.argument("TAXID", type=int)
@path_option
@click.pass_context
def update(ctx: Context, path: Path, taxid: int) -> None:
    """Update the specified OTU with new data."""
    repo = Repo(path)

    ctx.ensure_object(dict)
    ctx.obj = {
        "REPO": repo,
        "TAXID": taxid,
    }

    if not repo.get_otu_id_by_taxid(taxid):
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())


@update.command(name="automatic")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_auto_update(ctx: Context, ignore_cache: bool) -> None:
    """Automatically update an OTU with the latest data from NCBI."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        click.echo(f'Run "virtool otu create {taxid} --autofill" instead.')
        sys.exit(1)

    auto_update_otu(repo, otu_, ignore_cache=ignore_cache)


@update.command(name="promote")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_promote_accessions(
    ctx: Context,
    ignore_cache: bool,
) -> None:
    """Promote all RefSeq accessions within this OTU."""
    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(ctx.obj["TAXID"])

    promote_otu_accessions(repo, otu_, ignore_cache)


@update.command(name="isolate")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@click.pass_context
def isolate_create(
    ctx: Context,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
        except RefSeqConflictError as err:
            click.echo(
                f"{err.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as err:
        click.echo(
            f"{err.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)

    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@update.command(name="exclude")  # type: ignore
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.pass_context
def accession_exclude(
    ctx: Context,
    accessions_: list[str],
) -> None:
    """Exclude the given accessions from this OTU.

    Any duplicate accessions will be ignored. Only one exclusion per unique accession
    will be made.
    """
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    exclude_accessions_from_otu(repo, otu_, accessions_)


@update.command(name="default")  # type: ignore
@click.argument("ISOLATE_KEY", type=str)
@click.pass_context
def otu_set_representative_isolate(
    ctx: Context,
    isolate_key: str,
) -> None:
    """Update the OTU with a new representative isolate."""
    taxid = ctx.obj["TAXID"]

    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        sys.exit(1)

    try:
        isolate_id = UUID(isolate_key)
    except ValueError:
        parts = isolate_key.split(" ")
        try:
            isolate_name = IsolateName(IsolateNameType(parts[0].lower()), parts[1])
        except ValueError:
            click.echo(f'Error: "{isolate_key}" is not a valid isolate name.', err=True)
            sys.exit(1)

        isolate_id = otu_.get_isolate_id_by_name(isolate_name)

    if isolate_id is None:
        click.echo("Isolate could not be found in this OTU.", err=True)
        sys.exit(1)

    set_representative_isolate(repo, otu_, isolate_id)


@update.group()
@click.pass_context
def plan(ctx: Context) -> None:
    """Add to and replace isolate plans for this OTU."""


@plan.command(name="extend")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_extend_segment_list(
    ctx: Context,
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Add recommended or optional segments to the OTU plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        add_segments_to_plan(
            repo,
            otu_,
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)


@plan.command(name="multipartite")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--name",
    type=(str, str),
    help='a name for the originalsegment, e.g. "RNA A"',
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_resize_monopartite(
    ctx: Context,
    name: tuple[str, str],
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Change a monopartite OTU plan to a multipartite plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    prefix, key = name

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        resize_monopartite_plan(
            repo,
            otu_,
            name=SegmentName(prefix=prefix, key=key),
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
