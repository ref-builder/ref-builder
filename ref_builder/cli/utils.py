import sys
import warnings
from uuid import UUID

import click

from ref_builder.errors import InvalidInputError, PartialIDConflictError
from ref_builder.otu.builders.otu import OTUBuilder
from ref_builder.repo import Repo
from ref_builder.utils import OTUDeletedWarning

pass_repo = click.make_pass_decorator(Repo)

UUID_STRING_LENGTH = 36


def get_otu_from_identifier(repo: Repo, identifier: str) -> OTUBuilder:
    """Return an OTU id from the repo if identifier matches a single OTU.

    The identifier can either be a stringified UUIDv4, a truncated portion
    of a UUID, a NCBI Taxonomy ID or an acronym associated with the OTU.

    Raise a PartialIDConflictError if >1 OTU is found.

    :param repo: the repository to be searched.
    :param identifier: a non-UUID identifier.
        Can be an integer Taxonomy ID, acronym or truncated partial UUID.
    :return: the UUID of the OTU or ``None``
    """
    otu_id = None

    if len(identifier) == UUID_STRING_LENGTH:
        try:
            otu_id = UUID(identifier)
        except ValueError:
            otu_id = None

    elif identifier.isnumeric():
        try:
            taxid = int(identifier)
        except ValueError:
            pass
        else:
            otu_id = repo.get_otu_id_by_taxid(taxid)

    else:
        if (otu_id := repo.get_otu_id_by_acronym(identifier)) is not None:
            return repo.get_otu(otu_id)

        try:
            otu_id = repo.get_otu_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError as e:
            click.echo(
                e,
                err=True,
            )

    if otu_id is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    with warnings.catch_warnings(
        category=OTUDeletedWarning, record=True
    ) as warning_list:
        otu_ = repo.get_otu(otu_id)

    if otu_ is None:
        if warning_list:
            click.echo("OTU has been deleted.", err=True)

        else:
            click.echo("OTU not found.", err=True)

        sys.exit(1)

    return otu_


def get_otu_isolate_ids_from_identifier(repo: Repo, identifier: str) -> (UUID, UUID):
    """Return an isolate ID from the repo if identifier matches a single isolate.

    Handles cases where the isolate cannot be found.
    """
    isolate_id = None

    if len(identifier) == 32:
        try:
            isolate_id = UUID(identifier)
        except ValueError:
            isolate_id = None

    else:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except PartialIDConflictError:
            click.echo(
                "Partial ID too short to narrow down results.",
                err=True,
            )

        except InvalidInputError:
            click.echo(
                "Partial ID segment must be at least 8 characters long.", err=True
            )
            sys.exit(1)

    if isolate_id is not None:
        if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is not None:
            if isolate_id in repo.get_otu(otu_id).isolate_ids:
                return otu_id, isolate_id

    click.echo("Isolate ID could not be found.", err=True)
    sys.exit(1)
