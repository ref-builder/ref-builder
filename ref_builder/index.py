"""An index for improving repository performance."""

import binascii
import datetime
import sqlite3
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path
from uuid import UUID

from ref_builder.events.base import EventMetadata
from ref_builder.models.otu import OTUMinimal
from ref_builder.otu.builders.otu import OTUBuilder


@dataclass
class EventIndexItem:
    """The event IDs associated with an OTU."""

    event_ids: list[int]
    """A list of event IDs"""

    otu_id: UUID
    """The OTU ID this cache index is associated with."""


@dataclass
class Snapshot:
    """A snapshot of an OTU at a specific event."""

    at_event: int
    """The event at which the snapshot was taken."""

    otu: OTUBuilder
    """The OTU that was snapshotted."""


def adapt_uuid(original_uuid: UUID) -> str:
    """Automatically convert UUID to SQLite-compatible string."""
    return str(original_uuid)


def adapt_datetime(original_datetime: datetime.datetime) -> str:
    """Automatically convert datetime to SQLite-compatible string."""
    return original_datetime.isoformat()


def convert_uuid(string_uuid: bytes) -> UUID:
    """Automatically convert stringified UUID back to UUID type."""
    return UUID(string_uuid.decode())


def convert_datetime(string_datetime: bytes) -> datetime.datetime:
    """Automatically convert stringified datetime back to datetime type."""
    return datetime.datetime.fromisoformat(string_datetime.decode())


sqlite3.register_adapter(UUID, adapt_uuid)
sqlite3.register_adapter(datetime.datetime, adapt_datetime)

sqlite3.register_converter("uuid", convert_uuid)
sqlite3.register_converter("datetime", convert_datetime)


class Index:
    """An index for rapid access to repository data."""

    def __init__(self, path: Path) -> None:
        """Initialize the index."""
        self.path = path
        """The path to the index file."""

        self.path.parent.mkdir(exist_ok=True, parents=True)

        self.con = sqlite3.connect(
            path,
            isolation_level=None,
            detect_types=sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES,
        )

        self.con.execute("PRAGMA journal_mode = WAL")
        self.con.execute("PRAGMA synchronous = NORMAL")

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS events (
                event_id INTEGER PRIMARY KEY,
                otu_id TEXT,
                timestamp TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS isolates (
                id TEXT PRIMARY KEY,
                name TEXT,
                otu_id TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otus (
                id TEXT PRIMARY KEY,
                acronym TEXT,
                at_event INTEGER,
                name TEXT,
                otu JSONB,
                taxid INTEGER
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS sequences (
                id TEXT PRIMARY KEY,
                crc TEXT,
                otu_id TEXT,
                sequence TEXT
            );
            """,
        )

        self.con.execute(
            """
            CREATE TABLE IF NOT EXISTS otu_updates (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                otu_id TEXT,
                timestamp_complete TEXT
            );
            """,
        )

        for table, column in [
            ("events", "otu_id"),
            ("isolates", "id"),
            ("otus", "id"),
            ("otus", "name"),
            ("otus", "taxid"),
            ("sequences", "otu_id"),
            ("sequences", "crc"),
            ("otu_updates", "otu_id"),
        ]:
            self.con.execute(
                f"""
                CREATE INDEX IF NOT EXISTS {table}_{column} ON {table} ({column});
                """,
            )

        self.con.commit()

    @property
    def otu_ids(self) -> set[UUID]:
        """A list of all OTUs tracked in the index."""
        return {
            row[0] for row in self.con.execute('SELECT id AS "id [uuid]" FROM otus')
        }

    def add_event_id(
        self,
        event_id: int,
        otu_id: UUID,
        timestamp: datetime.datetime,
    ) -> None:
        """Add a new event ID and associated it with a given OTU ID.

        This allows fast lookup of event IDs for a given OTU ID.

        :param event_id: the event ID to add
        :param otu_id: the OTU ID to add the event ID to
        :param timestamp: the timestamp of the event
        """
        self.con.execute(
            """
            INSERT INTO events (event_id, otu_id, timestamp)
            VALUES (?, ?, ?)
            """,
            (
                event_id,
                otu_id,
                timestamp,
            ),
        )

    def get_event_ids_by_otu_id(self, otu_id: UUID) -> EventIndexItem | None:
        """Get a list of event IDs for an OTU."""
        res = self.con.execute(
            """
            SELECT event_id
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (otu_id,),
        )

        event_ids = [row[0] for row in res]

        if event_ids:
            return EventIndexItem(
                event_ids,
                otu_id,
            )

        return None

    def get_first_timestamp_by_otu_id(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the earliest event associated with this ID."""
        cursor = self.con.execute(
            """
            SELECT timestamp
            AS "timestamp [datetime]"
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_latest_timestamp_by_otu_id(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the latest event associated with this ID."""
        cursor = self.con.execute(
            """
            SELECT timestamp
            AS "timestamp [datetime]"
            FROM events
            WHERE otu_id = ?
            ORDER BY event_id DESC
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_acronym(self, acronym: str) -> UUID | None:
        """Get an OTU ID by its acronym."""
        if not acronym:
            raise ValueError("Acronym must contain at least one character.")

        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE acronym = ?',
            (acronym,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_name(self, name: str) -> UUID | None:
        """Get an OTU ID by its name."""
        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE name = ?',
            (name,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_taxid(self, taxid: int) -> UUID | None:
        """Get an OTU ID by its taxonomy ID."""
        cursor = self.con.execute(
            'SELECT id AS "id [uuid]" FROM otus WHERE taxid = ?',
            (taxid,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def get_id_by_isolate_id(self, isolate_id: UUID) -> UUID | None:
        """Get an OTU ID from an isolate ID that belongs to it."""
        cursor = self.con.execute(
            'SELECT otu_id AS "otu_id [uuid]" FROM isolates WHERE id = ?',
            (isolate_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def delete_otu(self, otu_id: UUID) -> None:
        """Remove an OTU from the index.

        This happens rarely, so we just rewrite the whole index.

        :param otu_id: the ID of the OTU to remove.
        """
        self.con.execute(
            "DELETE FROM isolates WHERE otu_id = ?",
            (otu_id,),
        )

        self.con.execute(
            "DELETE FROM otus WHERE id = ?",
            (otu_id,),
        )

        self.con.execute(
            "DELETE FROM sequences WHERE otu_id = ?",
            (otu_id,),
        )

        self.con.commit()

    def iter_event_metadata(self) -> Iterator[EventMetadata]:
        """Iterate over event metadata."""
        rows = self.con.execute(
            "SELECT event_id, otu_id, timestamp FROM events ORDER BY event_id",
        ).fetchall()

        for row in rows:
            yield EventMetadata(
                id=row[0],
                otu_id=UUID(row[1]) if row[1] else None,
                timestamp=row[2],
            )

    def iter_minimal_otus(self) -> Iterator[OTUMinimal]:
        """Iterate over minimal representations of all OTUs in the index."""
        rows = self.con.execute(
            """
            SELECT acronym, id AS "id [uuid]", name, taxid
            FROM otus ORDER BY name
            """,
        ).fetchall()

        for row in rows:
            yield OTUMinimal(
                acronym=row[0],
                id=row[1],
                name=row[2],
                taxid=row[3],
            )

    def get_last_otu_update_timestamp(self, otu_id: UUID) -> datetime.datetime | None:
        """Get the timestamp of the last event in an update for an OTU."""
        cursor = self.con.execute(
            """
            SELECT timestamp_complete FROM otu_updates
            WHERE otu_id = ? ORDER BY id DESC
            """,
            (otu_id,),
        )

        if result := cursor.fetchone():
            return datetime.datetime.fromisoformat(result[0])

        return None

    def add_otu_update_history_entry(
        self, otu_id: UUID, timestamp: datetime.datetime
    ) -> int | None:
        """Write an entry into the OTU update history."""
        self.con.execute(
            """
            INSERT INTO
            otu_updates(otu_id, timestamp_complete)
            VALUES(?, ?)
            """,
            (
                otu_id,
                timestamp,
            ),
        )

        cursor = self.con.execute(
            "SELECT id FROM otu_updates WHERE otu_id = ? ORDER BY id DESC",
            (otu_id,),
        )

        if result := cursor.fetchone():
            return result[0]

        return None

    def load_snapshot(self, otu_id: UUID) -> Snapshot | None:
        """Load an OTU snapshot."""
        cursor = self.con.execute(
            "SELECT at_event, otu FROM otus WHERE id = ?",
            (otu_id,),
        )

        result = cursor.fetchone()

        if not result:
            return None

        at_event, otu_json = result

        otu = OTUBuilder.model_validate_json(otu_json)

        sequence_ids = [
            sequence.id for isolate in otu.isolates for sequence in isolate.sequences
        ]

        # Fetch all sequences in a single query
        placeholders = ",".join("?" for _ in sequence_ids)

        sequence_map = dict(
            self.con.execute(
                f"SELECT id, sequence FROM sequences WHERE id IN ({placeholders})",
                sequence_ids,
            ).fetchall(),
        )

        # Update the data structure and check for missing sequences
        missing_sequences = []

        for isolate in otu.isolates:
            for sequence in isolate.sequences:
                sequence_id = str(sequence.id)

                if sequence_id in sequence_map:
                    sequence.sequence = sequence_map[sequence_id]
                else:
                    missing_sequences.append(str(sequence_id))

        # Raise an error if any sequences are missing
        if missing_sequences:
            raise ValueError(f"Sequences not found: {', '.join(missing_sequences)}")

        return Snapshot(
            at_event=at_event,
            otu=OTUBuilder.model_validate(otu),
        )

    def upsert_otu(self, otu: OTUBuilder, at_event: int) -> None:
        """Update the index based on a list of OTUs.

        OTUs that aren't already in the index will be added, and those that are will be
        updated.

        Only sequences that have changed will be updated.
        """
        sequence_ids = {seq.id for isolate in otu.isolates for seq in isolate.sequences}

        placeholders = ",".join("?" for _ in sequence_ids)

        # Delete any sequences that are no longer in the OTU.
        self.con.execute(
            f"""
            DELETE FROM sequences WHERE otu_id = ? AND NOT id IN ({placeholders});
            """,
            (otu.id, *sequence_ids),
        )

        batch = []

        # Insert or update the isolates.
        for isolate in otu.isolates:
            self.con.execute(
                "INSERT OR REPLACE INTO isolates (id, name, otu_id) VALUES (?, ?, ?)",
                (
                    isolate.id,
                    str(isolate.name),
                    otu.id,
                ),
            )

            for sequence in isolate.sequences:
                crc = _calculate_crc32(sequence.sequence)
                batch.append(
                    (
                        sequence.id,
                        crc,
                        otu.id,
                        sequence.sequence,
                    ),
                )

        # Perform batch insert
        self.con.executemany(
            """
            INSERT OR REPLACE INTO sequences (id, crc, otu_id, sequence)
            VALUES (?, ?, ?, ?)
            """,
            batch,
        )

        # Update only if CRC has changed
        self.con.executemany(
            """
            UPDATE sequences
            SET crc = ?, otu_id = ?, sequence = ?
            WHERE id = ? AND crc != ?
            """,
            [(crc, otu_id, seq, seq_id, crc) for seq_id, crc, otu_id, seq in batch],
        )

        self.con.execute(
            """
            INSERT OR REPLACE INTO otus (id, acronym, at_event, name, otu, taxid)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                otu.id,
                otu.acronym,
                at_event,
                otu.name,
                otu.model_dump_json(),
                otu.taxid,
            ),
        )

        self.con.commit()

    def prune(self, event_id: int) -> None:
        """Rollback the index to a previous event.

        * Deletes all events after the given event ID.
        * Deletes all OTU snapshots that have ``at_event`` later than ``event_id``.

        Removed snapshots will be automatically updated next time an OTU is fetched from
        the repository.

        :param event_id: the event ID to rollback to
        """
        self.con.execute(
            """
            DELETE FROM events WHERE event_id > ?
            """,
            (event_id,),
        )

        self.con.execute(
            """
            DELETE FROM otus WHERE at_event > ?
            """,
            (event_id,),
        )

        self.con.commit()


def _calculate_crc32(sequence: str) -> str:
    """Calculate the CRC32 checksum of a sequence.

    :param sequence: the sequence to calculate the checksum of.
    :return: the CRC32 checksum as a hexadecimal string.

    """
    crc = binascii.crc32(sequence.encode("utf-8"))

    # Convert CRC as a hexadecimal string.
    return hex(crc & 0xFFFFFFFF)[2:].zfill(8)
