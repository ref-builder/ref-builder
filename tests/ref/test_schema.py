from uuid import UUID

import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.ncbi.models import NCBIGenbank
from ref_builder.models import MolType
from ref_builder.otu.update import create_schema_from_records
from ref_builder.otu.utils import get_multipartite_segment_name, parse_segment_name
from ref_builder.schema import SegmentName, OTUSchema


class MockNCBISource:
    def __init__(self, segment_name: str):
        self.segment = segment_name


class MockNCBIGenbank:
    def __init__(self, moltype: MolType, source: MockNCBISource):
        self.moltype = moltype
        self.source = source


@pytest.mark.parametrize(
    "accessions",
    [
        ["NC_024301"],
        [
            "NC_010314",
            "NC_010315",
            "NC_010316",
            "NC_010317",
            "NC_010318",
            "NC_010319",
        ],
    ],
)
def test_create_schema_from_records(
    accessions: list[str],
    scratch_ncbi_client,
    snapshot: SnapshotAssertion,
):
    records = scratch_ncbi_client.fetch_genbank_records(accessions)
    auto_schema = create_schema_from_records(records)

    assert type(auto_schema) is OTUSchema

    assert auto_schema.model_dump() == snapshot(exclude=props("id"))

    for segment in auto_schema.segments:
        assert type(segment.id) is UUID


class TestSegmentNameParser:
    @pytest.mark.parametrize(
        "expected_result, test_strings",
        [
            (SegmentName(prefix="DNA", key="A"), ["DNA A", "DNA_A", "DNA-A"]),
            (SegmentName(prefix="RNA", key="BN"), ["RNA BN", "RNA_BN", "RNA-BN"]),
            (SegmentName(prefix="DNA", key="U3"), ["DNA U3", "DNA U3", "DNA-U3"]),
        ],
    )
    def test_ok(self, expected_result: str, test_strings: list[str]):
        assert (
            parse_segment_name(test_strings[0])
            ==
            parse_segment_name(test_strings[1])
            ==
            parse_segment_name(test_strings[2])
            ==
            expected_result
        )

    @pytest.mark.parametrize(
        "fail_case",
        ["", "*V/", "51f9a0bc-7b3b-434f-bf4c-f7abaa015b8d"]
    )
    def test_fail(self, fail_case: str):
        with pytest.raises(ValueError):
            parse_segment_name(fail_case)

    @pytest.mark.parametrize(
        "expected_result, test_name, test_moltype",
        [
            (SegmentName(prefix="DNA", key="A"), "A", MolType.DNA),
            (SegmentName(prefix="RNA", key="BN"), "BN", MolType.RNA),
            (SegmentName(prefix="DNA", key="U3"), "U3", MolType.DNA),
        ],
    )
    def test_parse_from_record_ok(self, expected_result, test_name, test_moltype, mocker):
        mock_record = MockNCBIGenbank(moltype=test_moltype, source=MockNCBISource(test_name))

        mocker.patch.object(NCBIGenbank, "__new__", return_value=mock_record)

        record = NCBIGenbank()

        assert get_multipartite_segment_name(record) == expected_result

    @pytest.mark.parametrize(
        "test_name, test_moltype",
        [
            ("", MolType.DNA),
            ("V#", MolType.RNA),
            ("51f9a0bc-7b3b-434f-bf4c-f7abaa015b8d", MolType.DNA)],
    )
    def test_parse_from_record_fail(self, test_name, test_moltype, mocker):
        mock_record = MockNCBIGenbank(moltype=test_moltype, source=MockNCBISource(test_name))

        mocker.patch.object(NCBIGenbank, "__new__", return_value=mock_record)

        record = NCBIGenbank()

        with pytest.raises(ValueError):
            get_multipartite_segment_name(record)
