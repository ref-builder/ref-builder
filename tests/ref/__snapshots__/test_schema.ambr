# serializer version: 1
# name: TestIsolatePlan.test_serialize_plan[accessions0-MonopartitePlan]
  dict({
    'length': 7424,
    'name': None,
  })
# ---
# name: TestIsolatePlan.test_serialize_plan[accessions1-MultipartitePlan]
  dict({
    'segments': list([
      dict({
        'length': 1090,
        'name': dict({
          'key': 'N',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1057,
        'name': dict({
          'key': 'U3',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1087,
        'name': dict({
          'key': 'S',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1074,
        'name': dict({
          'key': 'M',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1015,
        'name': dict({
          'key': 'C',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
      dict({
        'length': 1099,
        'name': dict({
          'key': 'R',
          'prefix': 'DNA',
        }),
        'required': 'required',
      }),
    ]),
  })
# ---
